payments/models.py

```diff
@@ -135,6 +135,71 @@ def get_success_url(self) -> str:
     def get_process_url(self) -> str:
         return reverse('process_payment', kwargs={'token': self.token})
 
+    def get_payment_url(self):
+        """
+        Get the url the view that handles the payment (payment_details() in documentation)
+        For now used only by PayU provider to redirect users back to CVV2 form
+        """
+        raise NotImplementedError()
+
+    def get_user(self):
+        """ Get the user asociated with this payment """
+        raise NotImplementedError()
+
+    def get_user_email(self):
+        """ Get user email """
+        try:
+            # Most common implementation
+            return self.get_user().email
+        except AttributeError:
+            raise NotImplementedError()
+
+    def get_user_first_name(self):
+        """
+        Get user first name for purposes of the payment provider
+        Used only by PayU provider for now
+        """
+        try:
+            # Most common implementation
+            return self.get_user().first_name
+        except AttributeError:
+            raise NotImplementedError()
+
+    def get_user_last_name(self):
+        """
+        Get user last name for purposes of the payment provider
+        Used only by PayU provider for now
+        """
+        try:
+            # Most common implementation
+            return self.get_user().last_name
+        except AttributeError:
+            raise NotImplementedError()
+
+    def get_renew_token(self):
```

Comment:
Since this needs to be implemented by an application developer: what's expected here?
Isn't this something internal to the provider?

Reply:
To make recurring payments work, there needs to be token associate with the user.
I wanted to provide standardized way to store textual tokens for usage across `django-payments` payment providers.

Reply:
You can use `payment.attrs` to store custom per-provider data. I'd suggest using that.

For now it's a proxy that JSON-encodes into a string field (legacy reasons). I'm planning to migrate this into a plain `JSONField`, but usage should remain pretty much the same.

Reply:
An alternative would also be for you to create an explicit `Subscription` object, and save the data there.

You can then link payments to subscriptions. I think this also slightly mirrors how payment providers themselves model these things.

Reply:
The `Subscription` model is similar to what I am storing the parameters of this method to:
https://github.com/django-getpaid/django-plans/blob/master/plans/models.py#L402

So the abstraction could look something like this:
```python
class BaseSubscription(models.Model):
    user_plan = models.OneToOneField('UserPlan', on_delete=models.CASCADE, related_name='recurring')
    token = models.CharField(
        _('recurring token'),
        help_text=_('Token, that will be used for payment renewal. Depends on used payment provider'),
        max_length=255,
        default=None,
        null=True,
        blank=True,
    )
    card_expire_year = models.IntegerField(null=True, blank=True)
    card_expire_month = models.IntegerField(null=True, blank=True)
    card_masked_number = models.CharField(null=True, blank=True, max_length=255)
        
    class Meta: 
        abstract = True

```

But I am not sure, how to implement this to `django-payments` without implying too much about the usage logic e.g. how to initialise such objects inside `Payment` model.

Reply:
Probably the concrete `Payment` class created by users of `django-plans` would have to include a field specific for it. You can also provide a mixin, so that implementors do something like

```python
class Payment(BasePayment, SubscriptionMixin):
    # SubscriptionMixin has a foreign key to the concrete subscription model.
```

I do have to point out that `card_expire_year`, `card_expire_month` and `card_masked_number` all make the assumption that the payment method was a credit card, which is not always the case.